### DISCUSSION TITLE
How to access roles from API Extension?

### DISCUSSION THREAD

**Discussion by rokdd** [directus/directus#14656] [Archive]
I am developing an API extension and I wanted to make sure that some roles are only to edit certain stuff. I managed so far to use the permission service with a hardcoded ID of the role. But how I can create or list all roles? In the api it is done very good, but in Javascript I do not see any option to read roles: https://github.com/directus/directus/blob/main/api/src/services/roles.ts ?

So what I am doing / thinking wrong here?

**Comment by br41nslug** [comment]
For API extensions using the Services is the way to go here. For example fetching all roles can be done like this (hook used as example):

```js
export default ({ action }, { services, exceptions, database: knex }) => {
    const { RolesService } = services;

    // Sync with other collection
    action('COLLECTION_A.items.update', async ({ collection, payload }, { schema }) => {
        // this request uses no permissions which can be unsafe!
        const roleService = new RolesService({ knex, schema });

        // accepts all the api parameters in json form like filter,fields,deep
        const allRoles = await roleService.readByQuery({
          limit: -1
        });
    });
}
```

**Reply by rokdd** [reply]
Well I was not aware of the superclass :-( but thanks you it works now:

```js
let roles_service = await new services.RolesService({ schema: req.schema, accountability: req.accountability });
```
